staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go

```diff
@@ -67,17 +73,56 @@ var (
 		},
 		[]string{"transformation_type"},
 	)
+
+	// keyIDTotal is the number of times a keyID is used
+	// e.g. apiserver_envelope_encryption_keyid_total counter
+	// apiserver_envelope_encryption_keyid_total{key_id="sha256",
+	// provider_name="providerName",transformation_type="from_storage"} 1
+	KeyIDTotal = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "keyid_total",
+			Help:           "Number of times a keyID is used split by transformation type and provider.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"transformation_type", "observer_type", "provider_name", "key_id"},
```

Comment:
what's the cardinality of `key_id` and why is it sometimes separated by an underscore and sometimes not?

Reply:
`key_id` is the label name, the actual value is a sha64 hash set by the kmsv2 envelop transformer.

Re: cardinality, the usage pattern suggests limited number of keyid because of the need for storage migration when the keyid rotates and usually there's only 2 keyids in use during the rotation.

Reply:
However in theory plugin authors _could_ enable frequent keyID creation and rotation such that we could end up with many keyIDs. Given that this metric is used for debugging and automation purpose, we don't want to remove the label. what would you recommend here to ensure we can handle in the case of unbounded number of keyIDs? Do we have any precedence for how to handle metrics like this?

Reply:
Unbounded labels are basically memory leaks and frequent rotation could effectively cause the apiserver to OOM. I would remove the label and use logs for this type of dimension.

Reply:
I disagree.  Being able to observe the key ID hash easily from the API server is incredibly valuable for the operator (to understand rotation and the state of data in etcd).  The plugin is completely under the control of the operator, there is no real risk of OOM here.

Reply:
Remember that key ID changes are both expensive for the plugin (it has to be able to decrypt all data from all old keys) and the cluster itself (you have to run storage migration to migrate old data to the new key).  This makes it impractical to do frequently (i.e. every few weeks is the most aggressive rotation interval).

If there is still concern, we can try to prune metrics after some N key ID hashes have been observed.

Reply:
> Being able to observe the key ID hash easily from the API server is incredibly valuable for the operator (to understand rotation and the state of data in etcd). 

You can also get this information (easily) from logs.

> Being able to observe the key ID hash easily from the API server is incredibly valuable for the operator (to understand rotation and the state of data in etcd). 

Again, best practice is to inject **logs** with dimensions which are unbounded and to use basically enums for values for labels in metrics. We've had many, many issues (https://github.com/kubernetes/kubernetes/issues/53485, https://github.com/kubernetes/kubernetes/issues/52121, https://github.com/kubernetes/kubernetes/pull/54921, etc.) in Kubernetes due to unbounded label growth and this is a pretty hard stance we take when looking at any new or recently modified metric. Even innocuous cases that look like this have resulted in situations which warranted reverting changes, and copious backports of fixes.

Reply:
Logs won't accumulate and persist state, which is one of the nice features of metrics.  So no, an automated process cannot easily get this information from them.

Let's fix the unbounded label growth instead of denying useful information to end users.  Two ideas come to mind: pruning old metrics based on a LRU cache or limiting the cardinality but hashing down to a smaller bit size.  Thoughts?

Reply:
> Logs won't accumulate and persist state

What? You can literally create metrics from logs. GCP has log-based metrics which is based on this principle. 

> pruning old metrics based on a LRU cache or limiting the cardinality but hashing down to a smaller bit size.

This is a counter metric, so pruning off an LRU cache will make the metric non-sensical. Hashing to a smaller bit size would work, but you'd need to guarantee it's sufficiently small and the downside would be that you'd be more prone to collisions. 

An approach that may actually work is if you keep a small number of the hashes in memory and just reset the metric when the number of hashes in memory exceed a certain number (you would also need to prune the set of hashes in memory when this happens) This will prevent the metric from growing past a certain size while leaving the meaning of the metric intact. 

But this is pretty complicated, I think it would be simpler to omit this dimension from the metric and log out the hash instead.

/cc @kubernetes/sig-instrumentation-approvers

Reply:
> > Logs won't accumulate and persist state
> 
> What? You can literally create metrics from logs. GCP has log-based metrics which is based on this principle.

Sorry, I was not clear enough.  I meant "by default without needing an external system to coordinate."  Anything is possible if you write enough code and have enough external systems/state.  If we include this key ID hash in the metric, a simple restart of the API server would tell me all key IDs in use (as the API server fills out its watch cache).  I do not think I can overstate how valuable this information is.

> > pruning old metrics based on a LRU cache or limiting the cardinality but hashing down to a smaller bit size.
> 
> This is a counter metric, so pruning off an LRU cache will make the metric non-sensical.

I meant that we would drop an entire set of metrics related to a particular key ID as it went off the cache.

> An approach that may actually work is if you keep a small number of the hashes in memory and just reset the metric when the number of hashes in memory exceed a certain number (you would also need to prune the set of hashes in memory when this happens) This will prevent the metric from growing past a certain size while leaving the meaning of the metric intact.

This sounds fine.  What is an appropriate number to keep around?

> But this is pretty complicated, I think it would be simpler to omit this dimension from the metric and log out the hash instead.

This is rather simple considering the complexity of the feature itself...

Reply:
> Sorry, I was not clear enough. I meant "by default without needing an external system to coordinate." Anything is possible if you write enough code and have enough external systems/state. If we include this key ID hash in the metric, a simple restart of the API server would tell me all key IDs in use (as the API server fills out its watch cache). I do not think I can overstate how valuable this information is.

Using metrics as an interface to a database because you have no other way to access that information doesn't make it a good idea.

> This sounds fine. What is an appropriate number to keep around?

A very small number, like 10, unless there is some compelling reason to keep more. 

> This is rather simple considering the complexity of the feature itself...

As a word of warning, I would not be onboard with promoting a metric augmented like this to a stability level that is greater than `ALPHA` due to nature of the modifications.

Reply:
> Using metrics as an interface to a database because you have no other way to access that information doesn't make it a good idea.

Fair enough.

> A very small number, like 10, unless there is some compelling reason to keep more.

10 works 👍 

> As a word of warning, I would not be onboard with promoting a metric augmented like this to a stability level that is greater than `ALPHA` due to nature of the modifications.

This is fine.  We have plans for a more robust rotation story as a pre-req for GA for KMS v2.

Reply:
The unbounded growth of the number of metrics exposed is indeed quite worrying, but that is only the case at ingestion because due to the nature of the key id not changing very frequently, it shouldn't be hard on the monitoring platform storage system as long as we find a way to limit the number of key id exposed.

A simpler solution to the various cache suggestion mentioned above would be to create const metrics from the cache: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#NewConstMetric

The idea is to keep the cache up-to-date with the latest key-id in use and prune the key-id that are not used anymore (post-migration). Every metrics endpoint scrape you would then generate NewConstMetric from all the data stored in the cache. This would result in only exposing information about the latest keys and pruning all the old ones that are not used anymore.

> This sounds fine. What is an appropriate number to keep around?

Wouldn't only 2 per transformer be needed at all time? During rotation only the last 2 hashes are in use and when the storage migration is complete only the new hash remain useful.
Even without storage migration in place, is it really useful for end users to know that the old KEK is still used for decryption?

Reply:
> Knowing that the key is still in-use with the timestamp metric should be enough to know that it shouldn't be deleted.

@dgrisonnet Since the `key_id_hash_last_timestamp_seconds` metric still includes the key_id label, that could also be unbounded.

Reply:
Correct, sorry about that, I got confused.

Assuming the metrics need to be added before storage migration is implemented there isn't much room to maneuver if we want to reliably know if a key is still being used. I would be fine with adding the metrics as is for now considering that they have a cardinality of 4 timeseries per key which sounds reasonable. There is still the problem of unbounded growth but considering how infrequently new keys are being added, it should be fine. Worst case scenario we have mechanisms to remove this label at runtime: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/2305-metrics-cardinality-enforcement.

But I would agree with @logicalhan that we shouldn't graduate these metrics to beta/stable unless there is a mechanism in place to bound them.

Reply:
Added pruning for key ID metrics to store last 10 `providerName + keyIDHash`. @logicalhan @enj PTAL